lib/srv/db/opensearch/engine.go

```diff
@@ -0,0 +1,283 @@
+		statusCode = http.StatusUnauthorized
+		cause.Type = "access_denied_exception"
+	}
+
+	jsonBody, err := json.Marshal(cause)
+	if err != nil {
+		e.Log.WithError(err).Error("failed to marshal error response")
+		return
+	}
+
+	response := &http.Response{
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: statusCode,
+		Body:       io.NopCloser(bytes.NewBuffer(jsonBody)),
+		Header: map[string][]string{
+			"Content-Type":   {"application/json"},
+			"Content-Length": {strconv.Itoa(len(jsonBody))},
+		},
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.Errorf("OpenSearch error: %+v", trace.Unwrap(err))
+		return
+	}
+}
+
+// HandleConnection authorizes the incoming client connection, connects to the
+// target OpenSearch server and starts proxying requests between client/server.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	err := e.checkAccess(ctx)
+	e.Audit.OnSessionStart(e.Context, e.sessionCtx, err)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := e.process(ctx, req); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// process reads request from connected OpenSearch client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) process(ctx context.Context, req *http.Request) error {
+	reqCopy, err := copyRequest(ctx, req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// force HTTPS, set host URL.
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = e.sessionCtx.Database.GetURI()
+
+	e.emitAuditEvent(reqCopy)
+
+	roleArn, err := libaws.BuildRoleARN(e.sessionCtx.DatabaseUser, e.sessionCtx.Database.GetAWS().Region, e.sessionCtx.Database.GetAWS().AccountID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	signCtx := &libaws.SigningCtx{
+		SigningName:   opensearchservice.EndpointsID,
+		SigningRegion: e.sessionCtx.Database.GetAWS().Region,
+		Expiry:        e.sessionCtx.Identity.Expires,
+		SessionName:   e.sessionCtx.Identity.Username,
+		AWSRoleArn:    roleArn,
+		AWSExternalID: e.sessionCtx.Database.GetAWS().ExternalID,
+	}
+
+	signedReq, err := e.signingSvc.SignRequest(e.Context, reqCopy, signCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, e.sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	clt := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}
+
+	// Send the request to elasticsearch API
+	resp, err := clt.Do(signedReq)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	return trace.Wrap(e.sendResponse(resp))
+}
+
+// copyRequest creates a shallow copy of the given request.
+func copyRequest(ctx context.Context, req *http.Request) (*http.Request, error) {
```

Comment:
I did this for Elasticsearch and OpenSearch. For others, the code is subtly different or uses a different body in cloned requests. We could _probably_ unify these cases, but I'd rather not shoehorn such changes into this PR. This would need to be extensively tested as we can get this subtly wrong.